File: Codes/Programmers/보석쇼핑.py
def solution(gems):
    n = len(gems)
    total_cnt = len(set(gems))
    result = []
    left, right = 0, 0
    counter = {gems[0]: 1}

    while left <= right < n:
        if len(counter) == total_cnt:
            result.append([left + 1, right + 1])
            counter[gems[left]] -= 1
            if counter[gems[left]] == 0:
                counter.pop(gems[left])
            left += 1
        else:
            right += 1
            if right < n:
                counter[gems[right]] = counter.get(gems[right], 0) + 1

    result.sort(key=lambda arr: (arr[1] - arr[0], arr[0]))
    return result[0]

File: Codes/Programmers/보석쇼핑.js
Map.prototype.getDefault = function (key, value) {
  return this.has(key) ? this.get(key) : value;
};

function solution(gems) {
  const allCount = [...new Set(gems)].length;
  const counter = new Map();
  counter.set(gems[0], 1);

  let minLength = gems.length + 1;
  let answer = [];

  let left = 0;
  let right = 0;
  while (right < gems.length) {
    if (counter.size === allCount) {
      const leftCount = counter.get(gems[left]);
      if (leftCount === 1) {
        counter.delete(gems[left]);
      } else {
        counter.set(gems[left], leftCount - 1);
      }
      if (minLength > right - left + 1) {
        answer = [left + 1, right + 1];
        minLength = right - left + 1;
      }
      left++;
    } else {
      right++;
      if (right >= gems.length) {
        break;
      }
      counter.set(gems[right], counter.getDefault(gems[right], 0) + 1);
    }
  }

  return answer;
}
